zebra-state/src/service/finalized_state/disk_format/upgrade/cache_genesis_roots.rs

```diff
@@ -0,0 +1,181 @@
+//! Updating the genesis note commitment trees to cache their roots.
+//!
+//! This reduces CPU usage when the genesis tree roots are used for transaction validation.
+//! Since mempool transactions are cheap to create, this is a potential remote denial of service.
+
+use std::sync::mpsc;
+
+use zebra_chain::{block::Height, sprout};
+
+use crate::service::finalized_state::{disk_db::DiskWriteBatch, ZebraDb};
+
+use super::CancelFormatChange;
+
+/// Runs disk format upgrade for changing the sprout and history tree key types.
+///
+/// Returns `Ok` if the upgrade completed, and `Err` if it was cancelled.
+///
+/// # Panics
+///
+/// If the state is empty.
+#[allow(clippy::unwrap_in_result)]
+#[instrument(skip(upgrade_db, cancel_receiver))]
+pub fn run(
+    initial_tip_height: Height,
+    upgrade_db: &ZebraDb,
+    cancel_receiver: &mpsc::Receiver<CancelFormatChange>,
+) -> Result<(), CancelFormatChange> {
+    let sprout_genesis_tree = sprout::tree::NoteCommitmentTree::default();
+    let sprout_tip_tree = upgrade_db.sprout_tree_for_tip();
+
+    let sapling_genesis_tree = upgrade_db
+        .sapling_tree_by_height(&Height(0))
+        .expect("caller has checked for genesis block");
+    let orchard_genesis_tree = upgrade_db
+        .orchard_tree_by_height(&Height(0))
+        .expect("caller has checked for genesis block");
+
+    // Writing the trees back to the database automatically caches their roots.
+    let mut batch = DiskWriteBatch::new();
+
+    // Fix the cached root of the Sprout genesis tree in its anchors column family.
+
+    // It's ok to write the genesis tree to the tip tree index, because is overwritten by
+    // the actual tip before the batch is written to the database.
+    batch.update_sprout_tree(upgrade_db, &sprout_genesis_tree);
```

Comment:
It's because of this line: https://github.com/zcashfoundation/zebra/blob/af5f97d43400be33a1ea725c1cd3d40a229f4dfa/zebra-state/src/service/finalized_state/zebra_db/shielded.rs#L645. 

We need to store the genesis tree at its anchor (together with its cached root). Note that anchors are the tree's roots. We then store the tip's tree again at another anchor. Since the roots of these two trees differ, the batch preserves both trees at their corresponding anchors.

Reply:
It's because of this line: https://github.com/zcashfoundation/zebra/blob/af5f97d43400be33a1ea725c1cd3d40a229f4dfa/zebra-state/src/service/finalized_state/zebra_db/shielded.rs#L645.

We store the genesis and tip trees at their corresponding anchors. Since the anchors differ (note that an anchor is the tree's root), the batch keeps both trees.

Reply:
> Why insert the genesis tree first if `sprout_tree_for_tip()` returns the default tree when empty and the value is always overwritten?

Also this covers the case where the tip is the genesis tree.